core/contacts/contacts.py

```diff
@@ -0,0 +1,293 @@
+                and pronunciations[full_name] != pronunciation
+            ):
+                logging.info(
+                    f"Multiple pronunciations found for '{full_name}'; using '{pronunciation}'"
+                )
+            pronunciations[full_name] = pronunciation
+
+            # Add pronunciation for each component of the name.
+            pron_parts = pronunciation.split()
+            name_parts = full_name.split()
+            if len(pron_parts) == len(name_parts):
+                for pron, name in zip(pron_parts, name_parts):
+                    if name in pronunciations and pronunciations[name] != pron:
+                        logging.info(
+                            f"Multiple different pronunciations found for '{name}' in "
+                            f"{full_name_raw}; using '{pron}'"
+                        )
+                    pronunciations[name] = pron
+            else:
+                logging.info(
+                    f"Pronunciation parts don't match name parts for '{full_name_raw}; skipping them.'"
+                )
+        else:
+            full_name = full_name_raw
+
+        # Handle nicknames with potential pronunciations
+        nicknames = []
+        for nickname_raw in contact.get("nicknames", []):
+            if ":" in nickname_raw:
+                pronunciation, nickname = [
+                    x.strip() for x in nickname_raw.split(":", 1)
+                ]
+                if (
+                    nickname in pronunciations
+                    and pronunciations[nickname] != pronunciation
+                ):
+                    logging.info(
+                        f"Multiple different pronunciations found for '{nickname}' in "
+                        f"contact {email}; using '{pronunciation}'"
+                    )
+                pronunciations[nickname] = pronunciation
+                nicknames.append(nickname)
+            else:
+                nicknames.append(nickname_raw)
+
+        return Contact(
+            email=email,
+            full_name=full_name,
+            nicknames=nicknames,
+            pronunciations=pronunciations,
+        )
+
+
+csv_contacts: list[Contact] = []
+json_contacts: list[Contact] = []
+
+
+@track_csv_list("contacts.csv", headers=("Name", "Email"), default={})
+def on_contacts_csv(values):
+    global csv_contacts
+    csv_contacts = []
+    for email, full_name in values.items():
+        if not email:
+            logging.error(f"Skipping contact missing email: {full_name}")
+            continue
+        csv_contacts.append(
+            Contact(email=email, full_name=full_name, nicknames=[], pronunciations={})
+        )
+    reload_contacts()
+
+
+@track_file("contacts.json", default="[]")
+def on_contacts_json(f):
+    global json_contacts
+    try:
+        contacts = json.load(f)
+    except Exception:
+        logging.exception("Error parsing contacts.json")
+        return
+
+    json_contacts = []
```

Comment:
(I'm probably missing something from the parsing logic above that makes from_json necessary).  

but since the contacts are data class, could just do
```
    json_contacts = [Contact(**contact) for contact in contacts]
```

Reply:
Oh wait, nevermind there's a dictionary in there :)